agent/system_metrics.go

```diff
@@ -0,0 +1,256 @@
+func (s *SystemMetricsCollector) Start(ctx context.Context) error {
+	var err error
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		return errors.Wrap(err, "problem creating cedar system metrics metadata object")
+	}
+
+	var streamingCtx context.Context
+	streamingCtx, s.streamingCancel = context.WithCancel(ctx)
+
+	for _, collector := range s.collectors {
+		stream, err := s.client.StreamSystemMetrics(streamingCtx, s.id, collector.Name(), metrics.StreamOpts{})
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem creating system metrics stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
+		err = ftdcCollector.SetMetadata(collector.Metadata())
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem setting metadata on stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		s.wg.Add(1)
+		go s.timedCollect(streamingCtx, collector, stream, ftdcCollector)
+	}
+	go s.closeOnCancel(ctx, streamingCtx)
+
+	return nil
+}
+
+func (s *SystemMetricsCollector) timedCollect(ctx context.Context, mc MetricCollector, stream *metrics.SystemMetricsWriteCloser, collector ftdc.Collector) {
+	defer s.wg.Done()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	defer s.catcher.Add(errors.Wrap(stream.Close(), fmt.Sprintf("problem closing system metrics stream for id %s and metricType %s", s.id, mc.Name())))
+	defer s.catcher.Add(errors.Wrap(ftdc.FlushCollector(collector, stream), fmt.Sprintf("problem flushing system metrics data collector for id %s and metricType %s", s.id, mc.Name())))
+	defer s.catcher.Add(errors.Wrap(recovery.HandlePanicWithError(recover(), nil, fmt.Sprintf("panic in system metrics stream for id %s and metricType %s", s.id, mc.Name())), ""))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			data, err := mc.Collect()
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem collecting system metrics data for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			err = collector.Add(data)
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem adding system metrics data to collector for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			_ = timer.Reset(s.interval)
+		}
+	}
+}
+
+func (s *SystemMetricsCollector) closeOnCancel(outerCtx, streamingCtx context.Context) {
+	for {
+		select {
+		case <-outerCtx.Done():
+			grip.Error(s.cleanup())
+			return
+		case <-streamingCtx.Done():
+			return
+		}
+	}
+}
+
+// Close cleans up any remaining connections and closes out the cedar
+// metadata if one was created with the completed_at timestamp and an
+// exit code. Close needs to be called before the context provided to
+// Start is cancelled.
+func (s *SystemMetricsCollector) Close() error {
```

Comment:
I was envisioning this would just cancel the streaming context, wait for all the goroutines to finish (including `closeOnCancel`) with `wg.Wait()`, and then return the errors after all the goroutines have stopped.